File: matlab/ex3/lrCostFunction.m
function [Jcost,grad] = lrCostFunction(X,Y,theta,lambda)
%% 函数功能：计算代价函数和梯度值
[m,~]=size(X);
hypthesis = sigmoid(X * theta);

%有正则项的代价函数表达式
Jcost = ( -Y.' * log(hypthesis) - (1-Y.') * log(1-hypthesis) ) / (m) + lambda * (theta.' * theta)  / (2*m) ;

grad =( (X.') * (hypthesis - Y) +lambda * [0;theta(2:end)] ) / m;

end

File: python/ex3/lrCostFunction.py
import numpy as np
from sigmoid import sigmoid
# 这个函数是用来求代价和梯度，和前面ex2的costfunction相同
def lr_cost_function(X,Y,theta,lmd):
	m = X.shape[0]
	g = sigmoid(X.dot(theta))
	# g.shape = [m,]
	# Y.shape = [m,1]
	# 两者相减得到的是[m,m]
	Y = Y.reshape(Y.size)

	cost = (-Y.T).dot(np.log(g)) - ((1-Y).T).dot(np.log(1-g))
	cost = cost /(m) + lmd * (theta.T).dot(theta) / (2*m)
	
	grad = (X.T).dot(g-Y)/ m
	grad[0] = grad[0]
	grad[1:] = grad[1:] + (lmd * theta[1:])/m

	return cost,grad